{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare stats of each captured region  Kronos\n",
    "\n",
    "\n",
    "##  Input file\n",
    "\n",
    "\n",
    "The input file for this notebook is a VCF file with the following characteristics:\n",
    "\n",
    " * **```ID```** column is in the format ```KronosXXXX.chrYY.NNNNNN``` where \n",
    "     * ```XXX``` is the ID of the mutant line\n",
    "     * ```YY``` is the chromosome where the mutation is\n",
    "     * ```NNNNNNN``` is the position of the mutation, based on the RefSeq reference. \n",
    " * A single sample named **```sample```**, with at least the following fields. \n",
    "    * ```AD``` Allelic depths for the ref and alt alleles in the order listed.\n",
    "    * ```DP``` Approximate read depth (reads with MQ=255 or with bad mates are filtered)\n",
    "    \n",
    "\n",
    "## Call Het/Hom again\n",
    "\n",
    "Use this criteria: \n",
    "\n",
    " \n",
    "**HOMOZYGOUS**\n",
    "HomMinCov3: homozygous mutation present in all reads from an individual and detected at least three times\n",
    "```\n",
    "DP>=3\n",
    "REF=0\n",
    "ALT>=3\n",
    " ```\n",
    "and if ```ALT>=3```\n",
    "Kronos: ```ALT/DP >=85%```, then it is a HOM\n",
    "Cadenza: ```ALT/DP >=90%``` then it is a HOM\n",
    " \n",
    "**HETEROZYGOUS**\n",
    "HetMinCov5: mutation detected in at least five reads\n",
    "```DP>=5\n",
    "REF>=1\n",
    "ALT>=5\n",
    " ```\n",
    "AND to reduce the probability of calling sequencing errors as heterozygous mutations in regions of high coverage we also determined the minimum percent of mutant reads required for calling a heterozygous mutation.\n",
    "For Kronos: ```ALT/DP >=15% …. But <85%```\n",
    "For Cadenza: ```ALT/DP >=10% …. But <90%```\n",
    " \n",
    "If ```ALT/DP <15%``` in Kronos or <10% in Cadenza then they are called wildtype.\n",
    "\n",
    "\n",
    "## Output file\n",
    "\n",
    "The script produces a file with all the snps and one with each region covered with SNPs, to be used to identifiy the residual hetorogeneity in a later step. \n",
    "\n",
    "| Column | Description                                                                 |\n",
    "|--------|----------------------------------------------------------------------------|\n",
    "| CHROM  | Chromosome where the variation is                                          |\n",
    "| POS    | Initial position of the variation |\n",
    "| ID     | ID of the SNP, with same convention as described above (```KronosXXXX.chrYY.NNNNNN```) |\n",
    "| REF    | Reference allele |\n",
    "| ALT    | Alternative allele |\n",
    "| LINE   | Linr where the mutation occurs |\n",
    "| MUT    | Transition of the mutation, it is the same as ```REF>ALT``` |\n",
    "| EMS    | Is the mutation EMS induced? True when $ MUT \\in \\{ \"G>A\", \"C>T\",\"A>G\",\"T>C\" \\} $ |\n",
    "| HET    | If the mutation is htereozygos or homozygloys **before** applying the rules before |\n",
    "| DIST   | Distance from the previous variation. This is the value used to break blocks |\n",
    "| SNP_COUNT       | Variations located in the current region and line. We expected to have 1 at most per line |\n",
    "| region_length   | length of the current region |\n",
    "| mut_no          | Number of mutations in the current regions |\n",
    "| ems_per         | Percentage of the mutations that are actually EMS. We expect a $100\\%$ |\n",
    "| avg_mut_per_pos | Average mutations per position (across lines in the region). Ideally it shoyld be 1 |\n",
    "| mut_density     | Mutation density of the line in the current region.   $ mut\\_density= \\frac{100 * mut\\_no}{region\\_length} $ |\n",
    "| region_id       | Unique identifierd of the region  |\n",
    "|region_start    | Start of the region |\n",
    "|region_end      | End of the region  |\n",
    "|DP      | Coverage **D**e**p**th |\n",
    "|RO      | **R**eference allele **O**bservations |\n",
    "|AO      | **A**lternative allele **O**bservations |\n",
    "|SNP_INDEX    | $\\frac{AO}{DP}$ |\n",
    "|EMS_GT  |  **Hom**ozygous or **Het**erozygous, based on the rules described above |\n",
    "|MUT_QUAL  | Quality of the mutation. **HIGH** if the $ ( DP>3 \\land EMS\\_GT = \\text{\"Hom\"}) \\lor ( DP>5 \\land EMS\\_GT == \\text{\"Het\"} )$. **LOW** if the $DP$ is not enough  |\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "false"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'bio'\n",
    "require 'zlib'  \n",
    "require 'bio-vcf'\n",
    "require 'csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":to_h"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def print_current_block(out, block )\n",
    "  len = block.size\n",
    "  block.each do |vcf_line |\n",
    "    vcf_line.block_size = len\n",
    "    out.puts vcf_line.to_block_arr.join(\"\\t\")\n",
    "  end\n",
    "end\n",
    "\n",
    "class BioVcf::VcfRecord\n",
    "    attr_accessor :distance, :snp_count, :block_size, :het \n",
    "    @@ems_mut = [\"G>A\", \"C>T\",\"A>G\",\"T>C\"]\n",
    "    \n",
    "    def mut\n",
    "        @mut ||= \"#{self.ref}>#{self.alt.join(\",\")}\"\n",
    "    end\n",
    "    \n",
    "    def ems?\n",
    "        @is_ems ||= @@ems_mut.include?(self.mut) \n",
    "    end\n",
    "    \n",
    "    def line\n",
    "        @line ||= self.id.split(\".\")[0]\n",
    "    end\n",
    "  \n",
    "    def chrom_short\n",
    "        @chrom_short ||= self.chrom.gsub(\"chr\",\"\") \n",
    "    end\n",
    "    \n",
    "    def to_vcf\n",
    "        @fields.join(\"\\t\") \n",
    "    end\n",
    "    \n",
    "    def chrom_short!\n",
    "        @fields[0]=chrom_short\n",
    "        @chrom = chrom_short\n",
    "    end\n",
    "  \n",
    "    def to_block_arr\n",
    "        [self.chrom, self.pos, self.id, self.ref, self.alt, self.line,\n",
    "            self.mut,self.ems?,self.gt, self.distance, self.snp_count, self.block_size]\n",
    "    end\n",
    "  \n",
    "    def snp?\n",
    "        @snp ||= @fields[3].size == 1 and @fields[4].size == 1\n",
    "    end\n",
    "    \n",
    "    def gt\n",
    "        @gt ||= self.sample[\"sample\"].gt == \"1/1\" ? \"Hom\" : \"Het\"\n",
    "    end \n",
    "    \n",
    "    def ad\n",
    "       @ad ||=  self.sample[\"sample\"].ad\n",
    "    end\n",
    "    \n",
    "    def ao\n",
    "        ad[1..-1]\n",
    "    end\n",
    "    \n",
    "    def ro\n",
    "        ad[0]\n",
    "    end\n",
    "    \n",
    "    def dp\n",
    "        @dp ||= self.sample[\"sample\"].dp\n",
    "    end\n",
    "    \n",
    "    def snp_index\n",
    "        dp = self.sample[\"sample\"].dp\n",
    "        return ao.map{|a| a.to_f / dp }\n",
    "    end\n",
    "    \n",
    "    def mut_qual(min_het, max_het)\n",
    "        return @mut_qual if @mut_qual\n",
    "        if ems_gt(min_het, max_het) == \"Hom\"\n",
    "            if dp >= 3\n",
    "                @mut_qual = \"HIGH\"\n",
    "            elsif dp >= 2\n",
    "                @mut_qual = \"LOW\"\n",
    "            else\n",
    "               @mut_qual  = \"NIL\" \n",
    "            end\n",
    "        else\n",
    "            if dp >= 5\n",
    "                @mut_qual = \"HIGH\"\n",
    "            elsif dp >= 3\n",
    "                @mut_qual = \"LOW\"\n",
    "            else\n",
    "               @mut_qual  = \"NIL\" \n",
    "            end\n",
    "        end\n",
    "        return @mut_qual\n",
    "    end\n",
    "    \n",
    "    def ems_gt(min_het,max_het)    \n",
    "        return snp_index.map do |s_i|\n",
    "            case \n",
    "            when s_i <= min_het\n",
    "                return  \"Ref\"\n",
    "            when s_i <= max_het\n",
    "                return  \"Het\"\n",
    "            else\n",
    "                return \"Hom\"\n",
    "            end\n",
    "        end\n",
    "    end \n",
    "end\n",
    "\n",
    "\n",
    "class BioVcf::VCFfile\n",
    "    def initialize(file: \"\", is_gz: true, lines_to_ignore: [])\n",
    "        @file = file\n",
    "        @is_gz = is_gz\n",
    "        @lines_to_ignore = lines_to_ignore\n",
    "    end\n",
    "    \n",
    "    def parseVCFheader(head_line=\"\")\n",
    "        m=/##INFO=<ID=(.+),Number=(.+),Type=(.+),Description=\"(.+)\">/.match(head_line)\n",
    "        {:id=>m[1],:number=>m[2],:type=>m[3],:desc=>m[4]}\n",
    "    end\n",
    "    \n",
    "    def each_filter_lines\n",
    "        return enum_for(:each_filter_lines) unless block_given?\n",
    "        it = self.each\n",
    "        while true\n",
    "            rec = it.next\n",
    "            yield rec unless @lines_to_ignore.include? rec.line\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    def each\n",
    "        return enum_for(:each) unless block_given? \n",
    "        if @is_gz\n",
    "            infile = open(@file)\n",
    "            io = Zlib::GzipReader.new(infile) \n",
    "        else\n",
    "            io =  File.open(@file)\n",
    "        end \n",
    "        header = BioVcf::VcfHeader.new \n",
    "        io.each_line do |line|  \n",
    "            line.chomp!\n",
    "            #puts line\n",
    "            if line =~ /^##fileformat=/     \n",
    "                header.add(line)  \n",
    "                while true\n",
    "                    headerline = io.readline\n",
    "                    headerline.chomp!\n",
    "                    unless headerline =~ /^#/\n",
    "                        line = headerline\n",
    "                        break #end of header\n",
    "                    end\n",
    "                    header.add(headerline)\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            fields = BioVcf::VcfLine.parse(line)\n",
    "            rec    = BioVcf::VcfRecord.new(fields,header)\n",
    "            yield rec\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    def to_h\n",
    "        h = Hash.new\n",
    "        self.each do |vcf_line|\n",
    "            h[vcf_line.id] = vcf_line\n",
    "        end\n",
    "        h\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":print_current_region"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RegionLineStats = Struct.new(:mut_no, :ems_no, :avg_mut_per_pos, :mut_density)\n",
    "class SNPRegion\n",
    "    attr_accessor :id, :fixed\n",
    "    def initialize(id, fixed:false)\n",
    "        @block = []\n",
    "        @line_stats = Hash.new { |h, k| h[k] = RegionLineStats.new(0,0,0,0) }\n",
    "        @id = id\n",
    "        @fixed = fixed\n",
    "    end\n",
    "    \n",
    "    def <<(pos)\n",
    "        enrich_position(pos)\n",
    "        @block << pos\n",
    "    end\n",
    "    \n",
    "    def enrich_position(pos)\n",
    "        len = pos.size\n",
    "        pos.each do |vcf_line|\n",
    "            ls = @line_stats[vcf_line.line]\n",
    "            ls.mut_no += 1\n",
    "            ls.ems_no += 1 if vcf_line.ems?\n",
    "            vcf_line.snp_count = len\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    def range(start:0, len:0)\n",
    "        @range = [start, start + len]\n",
    "    end\n",
    "\n",
    "    def region_range\n",
    "        return @range if @fixed\n",
    "        range = [0,0]\n",
    "        range[0] = @block.first[0].pos if @block.first.size > 0\n",
    "        range[1] = @block.last[0].pos + 1  if @block.last.size > 0\n",
    "        range\n",
    "    end\n",
    "    \n",
    "    def len\n",
    "        r = region_range\n",
    "        r[1] - r[0]\n",
    "    end\n",
    "\n",
    "    def iterate_range_by_position\n",
    "        @block.each do |position |\n",
    "            position.each do |vcf_line|\n",
    "                yield vcf_line\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    def enrich_region\n",
    "        return if @block.size == 0 \n",
    "        range = region_range\n",
    "        dist = range[1] - range[0]\n",
    "        size_sum = 0\n",
    "        @block.each do |position |\n",
    "            size_sum += position.size\n",
    "        end\n",
    "        \n",
    "        iterate_range_by_position do |vcf_line|\n",
    "            vcf_line.block_size = dist\n",
    "        end\n",
    "        @line_stats.each_pair do |k,region_line|\n",
    "            region_line.mut_density = 100 * region_line.mut_no.to_f / len\n",
    "            region_line.avg_mut_per_pos = size_sum.to_f / @line_stats.size #    \\/ number of lines \n",
    "        end\n",
    "    end\n",
    "    \n",
    "    def print_current_region(out, min_het, max_het)\n",
    "        iterate_range_by_position do |vcf_line|\n",
    "            arr = vcf_line.to_block_arr\n",
    "            region_line = @line_stats[vcf_line.line]\n",
    "            arr << region_line.mut_no\n",
    "            arr << region_line.ems_no\n",
    "            arr << region_line.avg_mut_per_pos\n",
    "            arr << region_line.mut_density\n",
    "            arr << self.id\n",
    "            arr << self.region_range[0]\n",
    "            arr << self.region_range[1]\n",
    "            arr << vcf_line.dp\n",
    "            arr << vcf_line.ro\n",
    "            arr << vcf_line.ao.join(\",\")\n",
    "            arr << vcf_line.snp_index.join(\",\")\n",
    "            arr << vcf_line.ems_gt(min_het, max_het)\n",
    "            arr << vcf_line.mut_qual(min_het, max_het)\n",
    "            out.puts arr.join(\"\\t\")\n",
    "        end\n",
    "    end\n",
    "    \n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":find_rh_regions"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_rh_regions(vcf_file, out_prefix, vcf_extra: {} ,region_gap:0, window_size:10000, min_het: 0.15, max_het: 0.85)\n",
    "  out_file = \"#{out_prefix}.tsv.gz\"  \n",
    "  current_chr = \"\"\n",
    "  last_position = 0\n",
    "  i = 0\n",
    "  prev = 0\n",
    "  prev_dist = 0\n",
    "  decide_by = region_gap > 0 ? :region : :window\n",
    "  out = Zlib::GzipWriter.open(out_file) \n",
    "  \n",
    "  out.puts [\"CHROM\", \"POS\", \"ID\", \"REF\",\n",
    "        \"ALT\", \"LINE\",\"MUT\",\"EMS\",\"HET\",\"DIST\",\"SNP_COUNT\", \"region_length\",\n",
    "        \"mut_no\", \n",
    "        \"ems_per\", \n",
    "        \"avg_mut_per_pos\",\n",
    "        \"mut_density\", \"region_id\", \"region_start\", \"region_end\", \"DP\",\"RO\",\"AO\",\"SNP_INDEX\", \"EMS_GT\",\"MUT_QUAL\"].join(\"\\t\")\n",
    "  current_position = Array.new\n",
    "  reg_id = 0\n",
    "    \n",
    "  current_region = SNPRegion.new(reg_id, fixed: decide_by == :window )\n",
    "  current_region.range(start:0, len:window_size) if decide_by == :window\n",
    "  last_window = 0\n",
    "  vcf_file  = BioVcf::VCFfile.new(file:vcf_file )\n",
    "  i=0\n",
    "  seen = []\n",
    "  vcf_file.each do |rec|\n",
    "    seen << rec.id    \n",
    "    dist = rec.pos - prev\n",
    "    current_window = rec.pos/window_size\n",
    "    dist = rec.pos if dist < 0\n",
    "    rec.distance = dist\n",
    "    if dist != 0    \n",
    "      prev = rec.pos\n",
    "      current_region  << current_position\n",
    "      current_position = Array.new\n",
    "    end\n",
    "        \n",
    "    break_block = dist > region_gap if decide_by == :region\n",
    "    break_block = last_window != current_window if decide_by == :window\n",
    "        \n",
    "    if break_block\n",
    "      current_region.enrich_region\n",
    "      current_region.print_current_region(out, min_het, max_het)\n",
    "      reg_id += 1\n",
    "      current_region = SNPRegion.new(reg_id, fixed: decide_by == :window )\n",
    "      current_region.range(start:window_size*current_window, len:window_size) if decide_by == :window\n",
    "    end\n",
    "        \n",
    "    current_position << rec\n",
    "    last_window = current_window\n",
    "    \n",
    "  end\n",
    "  current_region.enrich_region\n",
    "  current_region.print_current_region(out, min_het, max_het)\n",
    "  out.close\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#<File:out/20190709_Kronos_snps_with_density_and_het_gap_60.tsv.gz (closed)>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_file=\"in/20190709_kronos_merge_nov2018_filter_incomplete.vcf.gz\"\n",
    "find_rh_regions(m_file, \"out/20190709_Kronos_snps_with_density_and_het_gap_60\",region_gap:60,min_het: 0.15, max_het: 0.85)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 2.5.1",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "2.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
